optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/descriptor/SolutionDescriptor.java

```diff
@@ -737,15 +740,18 @@ public boolean hasEntityDescriptor(Class<?> entitySubclass) {
     }
 
     public EntityDescriptor<Solution_> findEntityDescriptor(Class<?> entitySubclass) {
-        return lowestEntityDescriptorMemoization.computeIfAbsent(entitySubclass, key -> {
-            // Reverse order to find the nearest ancestor
-            for (Class<?> entityClass : reversedEntityClassList) {
-                if (entityClass.isAssignableFrom(entitySubclass)) {
-                    return entityDescriptorMap.get(entityClass);
-                }
-            }
-            return null;
-        });
+        EntityDescriptor<Solution_> foundEntityDescriptor =
+                lowestEntityDescriptorMemoization.computeIfAbsent(entitySubclass, key -> {
+                    // Reverse order to find the nearest ancestor
+                    for (Class<?> entityClass : reversedEntityClassList) {
+                        if (entityClass.isAssignableFrom(entitySubclass)) {
+                            return entityDescriptorMap.get(entityClass);
+                        }
+                    }
+                    // Dummy entity descriptor value instead of null, so that this properly works as a cache.
+                    return (EntityDescriptor<Solution_>) DUMMY_ENTITY_DESCRIPTOR;
+                });
+        return (foundEntityDescriptor == DUMMY_ENTITY_DESCRIPTOR) ? null : foundEntityDescriptor;
```

Comment:
This code now works as a cache even when the descriptor is not found. (= null.)
Therefore when this method is called in CS during `forEach(ProblemFact.class)`, it is significantly faster on average.